File: python/src/memosift/adapters/claude_agent_sdk.py
# Claude Agent SDK adapter — handles stateful sessions, compaction boundaries, hooks.
from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

from memosift.core.pipeline import compress
from memosift.core.types import (
    AnchorLedger,
    MemoSiftMessage,
    ToolCall,
    ToolCallFunction,
)
from memosift.providers.base import LLMResponse, MemoSiftLLMProvider

if TYPE_CHECKING:
    from memosift.config import MemoSiftConfig
    from memosift.report import CompressionReport


class ClaudeAgentLLMProvider:
    """Wraps a Claude Agent SDK client into MemoSiftLLMProvider.

    Uses the underlying Anthropic client for generation and token counting.
    """

    def __init__(self, client: Any, model: str = "claude-sonnet-4-6"):
        self._client = client
        self._model = model

    async def generate(
        self,
        prompt: str,
        max_tokens: int = 2048,
        temperature: float = 0.0,
    ) -> LLMResponse:
        """Generate text via the underlying Anthropic Messages API."""
        # The Agent SDK wraps an Anthropic client internally.
        anthropic_client = getattr(self._client, "_client", self._client)
        resp = await anthropic_client.messages.create(
            model=self._model,
            max_tokens=max_tokens,
            temperature=temperature,
            messages=[{"role": "user", "content": prompt}],
        )
        return LLMResponse(
            text=resp.content[0].text,
            input_tokens=resp.usage.input_tokens,
            output_tokens=resp.usage.output_tokens,
        )

    async def count_tokens(self, text: str) -> int:
        """Count tokens using heuristic (Agent SDK doesn't expose tokenizer)."""
        import math

        return math.ceil(len(text) / 3.5)


def adapt_in(messages: list[Any]) -> list[MemoSiftMessage]:
    """Convert Claude Agent SDK message objects to MemoSiftMessage list.

    Handles:
    - ``AssistantMessage`` with ``TextBlock`` and ``ToolUseBlock`` content
    - ``UserMessage`` with ``ToolResultBlock`` content
    - ``SystemMessage`` with ``subtype`` (init, compact_boundary)
    - Plain dicts (from ``get_session_messages()``)
    - Compaction boundaries are tagged as already compressed
    """
    result: list[MemoSiftMessage] = []

    for msg in messages:
        # Handle dict-style messages (from session history).
        if isinstance(msg, dict):
            result.extend(_adapt_dict_message(msg))
            continue

        # Handle typed Agent SDK message objects.
        msg_type = type(msg).__name__
        role = getattr(msg, "role", None)
        content_blocks = getattr(msg, "content", [])

        if msg_type == "SystemMessage":
            subtype = getattr(msg, "subtype", "")
            content = getattr(msg, "content", "")
            if isinstance(content, list):
                content = " ".join(getattr(b, "text", str(b)) for b in content)
            dm = MemoSiftMessage(
                role="system",
                content=content if isinstance(content, str) else str(content),
                metadata={"_agent_sdk_subtype": subtype},
            )
            # Mark compaction boundaries as already compressed.
            if subtype == "compact_boundary":
                dm._memosift_compressed = True
            result.append(dm)
            continue

        if msg_type == "ResultMessage":
            # Terminal message with metrics — skip (not conversation content).
            continue

        # Process content blocks (AssistantMessage, UserMessage).
        # Preserve original blocks for lossless round-trip reconstruction.
        original_blocks = list(content_blocks) if isinstance(content_blocks, list) else None

        text_parts: list[str] = []
        tool_calls: list[ToolCall] = []
        tool_results: list[tuple[str, str]] = []  # (tool_use_id, content)

        if isinstance(content_blocks, str):
            text_parts.append(content_blocks)
        elif isinstance(content_blocks, list):
            for block in content_blocks:
                block_type = getattr(block, "type", None) or (
                    block.get("type") if isinstance(block, dict) else None
                )
                if block_type == "thinking":
                    # Thinking blocks must not be compressed; skip them so they
                    # are preserved verbatim via _original_blocks round-trip.
                    continue
                elif block_type == "text":
                    text_parts.append(
                        getattr(block, "text", "") or block.get("text", "")
                        if isinstance(block, dict)
                        else getattr(block, "text", "")
                    )
                elif block_type == "tool_use":
                    tc_id = getattr(block, "id", "") or block.get("id", "")
                    tc_name = getattr(block, "name", "") or block.get("name", "")
                    tc_input = getattr(block, "input", {}) or block.get("input", {})
                    tool_calls.append(
                        ToolCall(
                            id=tc_id,
                            function=ToolCallFunction(
                                name=tc_name,
                                arguments=(
                                    json.dumps(tc_input)
                                    if isinstance(tc_input, dict)
                                    else str(tc_input)
                                ),
                            ),
                        )
                    )
                elif block_type == "tool_result":
                    tr_id = getattr(block, "tool_use_id", "") or block.get("tool_use_id", "")
                    tr_content = getattr(block, "content", "") or block.get("content", "")
                    if isinstance(tr_content, list):
                        tr_content = " ".join(
                            getattr(b, "text", str(b))
                            if not isinstance(b, dict)
                            else b.get("text", str(b))
                            for b in tr_content
                        )
                    tool_results.append((tr_id, str(tr_content)))
                elif isinstance(block, str):
                    text_parts.append(block)

        combined_text = "\n".join(text_parts) if text_parts else ""
        effective_role = role or ("assistant" if tool_calls else "user")

        result.append(
            MemoSiftMessage(
                role=effective_role,
                content=combined_text,
                tool_calls=tool_calls if tool_calls else None,
                metadata={
                    "_agent_sdk_type": msg_type,
                    "_original_blocks": original_blocks,
                    "_original_block_format": "claude_agent_sdk",
                },
            )
        )

        # Emit tool results as separate tool messages.
        for tool_use_id, tr_content in tool_results:
            result.append(
                MemoSiftMessage(
                    role="tool",
                    content=tr_content,
                    tool_call_id=tool_use_id,
                )
            )

    return result


def adapt_out(messages: list[MemoSiftMessage]) -> list[dict[str, Any]]:
    """Convert MemoSiftMessage list back to Anthropic-compatible format.

    Returns dicts suitable for the Anthropic Messages API, which the
    Agent SDK uses internally. Tool results are re-nested into user messages.
    """
    result: list[dict[str, Any]] = []

    tool_result_map: dict[str, MemoSiftMessage] = {}
    for msg in messages:
        if msg.role == "tool" and msg.tool_call_id:
            tool_result_map[msg.tool_call_id] = msg

    consumed_tool_ids: set[str] = set()

    for msg in messages:
        if msg.role == "tool":
            continue  # Re-nested below.

        blocks: list[dict[str, Any]] = []

        original_blocks = msg.metadata.get("_original_blocks")
        if original_blocks is not None:
            # Lossless round-trip: walk the original blocks and patch in
            # compressed content / surviving tool_calls.
            tc_map: dict[str, ToolCall] = {}
            if msg.tool_calls:
                for tc in msg.tool_calls:
                    tc_map[tc.id] = tc

            text_replaced = False
            for orig_block in original_blocks:
                # Handle both typed objects and dicts.
                btype = getattr(orig_block, "type", None) or (
                    orig_block.get("type") if isinstance(orig_block, dict) else None
                )

                if btype == "text":
                    if not text_replaced:
                        # First text block gets the (possibly compressed) content.
                        blocks.append({"type": "text", "text": msg.content})
                        text_replaced = True
                    # Subsequent text blocks are dropped — their content was
                    # merged into msg.content during adapt_in.

                elif btype == "thinking":
                    # Pass thinking blocks through unchanged.
                    if isinstance(orig_block, dict):
                        blocks.append(dict(orig_block))
                    else:
                        # Convert typed object to dict.
                        blocks.append(
                            {
                                "type": "thinking",
                                "thinking": getattr(orig_block, "thinking", ""),
                            }
                        )

                elif btype == "tool_use":
                    tc_id = getattr(orig_block, "id", "") or (
                        orig_block.get("id", "") if isinstance(orig_block, dict) else ""
                    )
                    if tc_id in tc_map:
                        tc = tc_map[tc_id]
                        try:
                            input_data = json.loads(tc.function.arguments)
                        except (json.JSONDecodeError, ValueError):
                            input_data = {"raw": tc.function.arguments}
                        blocks.append(
                            {
                                "type": "tool_use",
                                "id": tc.id,
                                "name": tc.function.name,
                                "input": input_data,
                            }
                        )

                else:
                    # Unknown / future block types — pass through unchanged.
                    if isinstance(orig_block, dict):
                        blocks.append(dict(orig_block))
                    elif isinstance(orig_block, str):
                        blocks.append(orig_block)
                    else:
                        # Convert typed object to dict representation.
                        blocks.append({"type": btype} if btype else orig_block)

            # If no text block existed in originals but we have content, add it.
            if not text_replaced and msg.content:
                blocks.insert(0, {"type": "text", "text": msg.content})
        else:
            # Fallback: reconstruct from scratch.
            if msg.content:
                blocks.append({"type": "text", "text": msg.content})

            if msg.role == "assistant" and msg.tool_calls:
                for tc in msg.tool_calls:
                    try:
                        input_data = json.loads(tc.function.arguments)
                    except (json.JSONDecodeError, ValueError):
                        input_data = {"raw": tc.function.arguments}
                    blocks.append(
                        {
                            "type": "tool_use",
                            "id": tc.id,
                            "name": tc.function.name,
                            "input": input_data,
                        }
                    )

        result.append({"role": msg.role, "content": blocks})

        # Re-nest tool results after assistant tool_use messages.
        if msg.role == "assistant" and msg.tool_calls:
            tr_blocks: list[dict[str, Any]] = []
            for tc in msg.tool_calls:
                if tc.id in tool_result_map and tc.id not in consumed_tool_ids:
                    tr_msg = tool_result_map[tc.id]
                    tr_blocks.append(
                        {
                            "type": "tool_result",
                            "tool_use_id": tc.id,
                            "content": tr_msg.content,
                        }
                    )
                    consumed_tool_ids.add(tc.id)
            if tr_blocks:
                result.append({"role": "user", "content": tr_blocks})

    return result


async def compress_agent_sdk_messages(
    messages: list[Any],
    *,
    llm: MemoSiftLLMProvider | None = None,
    config: MemoSiftConfig | None = None,
    task: str | None = None,
    ledger: AnchorLedger | None = None,
    client: Any = None,
    model: str = "claude-sonnet-4-6",
) -> tuple[list[dict[str, Any]], CompressionReport]:
    """Compress Claude Agent SDK session messages end-to-end.

    Respects compaction boundaries (``compact_boundary`` markers are
    tagged ``_memosift_compressed=True`` and skip re-compression).

    Args:
        messages: Agent SDK message objects or session history dicts.
        llm: MemoSiftLLMProvider (or auto-wrap ``client``).
        config: Pipeline configuration.
        task: Optional task description.
        ledger: Optional AnchorLedger.
        client: Claude Agent SDK client for auto-wrapping.
        model: Model name for auto-wrapping.

    Returns:
        Tuple of (compressed Anthropic-format messages, compression report).
    """
    provider = llm
    if provider is None and client is not None:
        provider = ClaudeAgentLLMProvider(client, model)

    memosift_msgs = adapt_in(messages)
    compressed, report = await compress(
        memosift_msgs,
        llm=provider,
        config=config,
        task=task,
        ledger=ledger,
    )
    return adapt_out(compressed), report


def _adapt_dict_message(msg: dict[str, Any]) -> list[MemoSiftMessage]:
    """Convert a plain dict message (from session history) to MemoSiftMessage(s).

    Returns a list because a single Anthropic-format user message with
    tool_result blocks expands into multiple MemoSiftMessages (one per result).
    """
    role = msg.get("role", "user")
    content = msg.get("content", "")

    if not isinstance(content, list):
        return [
            MemoSiftMessage(
                role=role,
                content=content if isinstance(content, str) else str(content),
            )
        ]

    # Preserve original blocks for lossless round-trip reconstruction.
    original_blocks = list(content)

    text_parts: list[str] = []
    tool_calls: list[ToolCall] = []
    tool_results: list[tuple[str, str]] = []  # (tool_use_id, content)

    for block in content:
        if isinstance(block, dict):
            if block.get("type") == "thinking":
                # Thinking blocks must not be compressed; skip them so they
                # are preserved verbatim via _original_blocks round-trip.
                continue
            elif block.get("type") == "text":
                text_parts.append(block.get("text", ""))
            elif block.get("type") == "tool_use":
                tool_calls.append(
                    ToolCall(
                        id=block.get("id", ""),
                        function=ToolCallFunction(
                            name=block.get("name", ""),
                            arguments=json.dumps(block.get("input", {})),
                        ),
                    )
                )
            elif block.get("type") == "tool_result":
                tr_content = block.get("content", "")
                if isinstance(tr_content, list):
                    tr_content = " ".join(
                        b.get("text", str(b)) for b in tr_content if isinstance(b, dict)
                    )
                tool_results.append((block.get("tool_use_id", ""), str(tr_content)))

    result: list[MemoSiftMessage] = []
    combined_text = "\n".join(text_parts) if text_parts else ""

    # Main message (assistant with tool_calls, or user with text).
    if combined_text or tool_calls:
        result.append(
            MemoSiftMessage(
                role=role,
                content=combined_text,
                tool_calls=tool_calls if tool_calls else None,
                metadata={
                    "_original_blocks": original_blocks,
                    "_original_block_format": "claude_agent_sdk",
                },
            )
        )

    # Emit tool results as separate tool messages.
    for tool_use_id, tr_content in tool_results:
        result.append(
            MemoSiftMessage(
                role="tool",
                content=tr_content,
                tool_call_id=tool_use_id,
            )
        )

    # If we produced nothing, emit an empty message to preserve the role.
    if not result:
        result.append(MemoSiftMessage(role=role, content=""))

    return result

File: typescript/src/adapters/claude-agent-sdk.ts
// Claude Agent SDK adapter — handles stateful sessions, compaction boundaries, hooks.

import type { MemoSiftConfig } from "../config.js";
import type { CompressResult } from "../core/pipeline.js";
import { compress } from "../core/pipeline.js";
import type { AnchorLedger, MemoSiftMessage, ToolCall } from "../core/types.js";
import { createMessage } from "../core/types.js";
import type { LLMResponse, MemoSiftLLMProvider } from "../providers/base.js";

/**
 * Wraps a Claude Agent SDK client into MemoSiftLLMProvider.
 *
 * Uses the underlying Anthropic client for generation and token counting.
 */
export class ClaudeAgentLLMProvider implements MemoSiftLLMProvider {
  private readonly _client: unknown;
  private readonly _model: string;

  constructor(client: unknown, model = "claude-sonnet-4-6") {
    this._client = client;
    this._model = model;
  }

  async generate(
    prompt: string,
    options?: { maxTokens?: number; temperature?: number },
  ): Promise<LLMResponse> {
    const maxTokens = options?.maxTokens ?? 2048;
    const temperature = options?.temperature ?? 0.0;

    // The Agent SDK wraps an Anthropic client internally.
    const anthropicClient = _getNestedProp(this._client, "_client") ?? this._client;
    const messages = anthropicClient as Record<string, unknown>;
    const messagesApi = _getNestedProp(messages, "messages") as Record<string, unknown>;

    const resp = (await (messagesApi.create as (args: unknown) => Promise<unknown>)({
      model: this._model,
      max_tokens: maxTokens,
      temperature,
      messages: [{ role: "user", content: prompt }],
    })) as {
      content: Array<{ text: string }>;
      usage: { input_tokens: number; output_tokens: number };
    };

    return {
      text: resp.content[0]!.text,
      inputTokens: resp.usage.input_tokens,
      outputTokens: resp.usage.output_tokens,
    };
  }

  async countTokens(text: string): Promise<number> {
    return Math.ceil(text.length / 3.5);
  }
}

// ── adaptIn ────────────────────────────────────────────────────────────────

/**
 * Convert Claude Agent SDK message objects to MemoSiftMessage list.
 *
 * Handles:
 * - AssistantMessage with TextBlock and ToolUseBlock content
 * - UserMessage with ToolResultBlock content
 * - SystemMessage with subtype (init, compact_boundary)
 * - Plain dicts (from get_session_messages())
 * - Compaction boundaries are tagged as already compressed
 */
export function adaptIn(messages: unknown[]): MemoSiftMessage[] {
  const result: MemoSiftMessage[] = [];

  for (const msg of messages) {
    // Handle dict-style messages (from session history).
    if (_isPlainObject(msg)) {
      result.push(..._adaptDictMessage(msg as Record<string, unknown>));
      continue;
    }

    // Handle typed Agent SDK message objects.
    const msgType = _getConstructorName(msg);
    const role = _getNestedProp(msg, "role") as string | null;
    const contentBlocks = _getNestedProp(msg, "content") ?? [];

    if (msgType === "SystemMessage") {
      const subtype = (_getNestedProp(msg, "subtype") as string) ?? "";
      let content = _getNestedProp(msg, "content");
      if (Array.isArray(content)) {
        content = content
          .map((b: unknown) =>
            typeof b === "string" ? b : ((_getNestedProp(b, "text") as string) ?? String(b)),
          )
          .join(" ");
      }
      const dm = createMessage(
        "system",
        typeof content === "string" ? content : String(content ?? ""),
        {
          metadata: { _agent_sdk_subtype: subtype },
        },
      );
      // Mark compaction boundaries as already compressed.
      if (subtype === "compact_boundary") {
        dm._memosiftCompressed = true;
      }
      result.push(dm);
      continue;
    }

    if (msgType === "ResultMessage") {
      // Terminal message with metrics — skip (not conversation content).
      continue;
    }

    // Process content blocks (AssistantMessage, UserMessage).
    // Preserve original blocks for lossless round-trip reconstruction.
    const originalBlocks: unknown[] | null = Array.isArray(contentBlocks)
      ? [...(contentBlocks as unknown[])]
      : null;

    const textParts: string[] = [];
    const toolCalls: ToolCall[] = [];
    const toolResults: Array<[string, string]> = [];

    if (typeof contentBlocks === "string") {
      textParts.push(contentBlocks);
    } else if (Array.isArray(contentBlocks)) {
      for (const block of contentBlocks as unknown[]) {
        const blockType = _getBlockType(block);

        if (blockType === "thinking") {
        } else if (blockType === "text") {
          textParts.push(_getBlockText(block));
        } else if (blockType === "tool_use") {
          const tcId = _getBlockProp(block, "id");
          const tcName = _getBlockProp(block, "name");
          const tcInput =
            _getNestedProp(block, "input") ??
            (_isPlainObject(block) ? (block as Record<string, unknown>).input : {});
          toolCalls.push({
            id: tcId,
            type: "function",
            function: {
              name: tcName,
              arguments: typeof tcInput === "object" ? JSON.stringify(tcInput) : String(tcInput),
            },
          });
        } else if (blockType === "tool_result") {
          const trId = _getBlockProp(block, "tool_use_id");
          let trContent =
            _getNestedProp(block, "content") ??
            (_isPlainObject(block) ? (block as Record<string, unknown>).content : "");
          if (Array.isArray(trContent)) {
            trContent = (trContent as unknown[])
              .map((b: unknown) => {
                if (_isPlainObject(b)) return (b as Record<string, unknown>).text ?? String(b);
                return (_getNestedProp(b, "text") as string) ?? String(b);
              })
              .join(" ");
          }
          toolResults.push([trId, String(trContent)]);
        } else if (typeof block === "string") {
          textParts.push(block);
        }
      }
    }

    const combinedText = textParts.length > 0 ? textParts.join("\n") : "";
    const effectiveRole = (role ??
      (toolCalls.length > 0 ? "assistant" : "user")) as MemoSiftMessage["role"];

    result.push(
      createMessage(effectiveRole, combinedText, {
        toolCalls: toolCalls.length > 0 ? toolCalls : null,
        metadata: {
          _agent_sdk_type: msgType,
          _original_blocks: originalBlocks,
          _original_block_format: "claude_agent_sdk",
        },
      }),
    );

    // Emit tool results as separate tool messages.
    for (const [toolUseId, trContent] of toolResults) {
      result.push(
        createMessage("tool", trContent, {
          toolCallId: toolUseId,
        }),
      );
    }
  }

  return result;
}

// ── adaptOut ───────────────────────────────────────────────────────────────

/**
 * Convert MemoSiftMessage list back to Anthropic-compatible format.
 *
 * Returns dicts suitable for the Anthropic Messages API, which the
 * Agent SDK uses internally. Tool results are re-nested into user messages.
 */
export function adaptOut(messages: MemoSiftMessage[]): Record<string, unknown>[] {
  const result: Record<string, unknown>[] = [];

  // Build tool result lookup.
  const toolResultMap = new Map<string, MemoSiftMessage>();
  for (const msg of messages) {
    if (msg.role === "tool" && msg.toolCallId) {
      toolResultMap.set(msg.toolCallId, msg);
    }
  }

  const consumedToolIds = new Set<string>();

  for (const msg of messages) {
    if (msg.role === "tool") {
      continue; // Re-nested below.
    }

    const blocks: Record<string, unknown>[] = [];
    const originalBlocks = msg.metadata._original_blocks as unknown[] | undefined;

    if (originalBlocks != null) {
      // Lossless round-trip: walk the original blocks and patch in
      // compressed content / surviving tool_calls.
      const tcMap = new Map<string, ToolCall>();
      if (msg.toolCalls) {
        for (const tc of msg.toolCalls) {
          tcMap.set(tc.id, tc);
        }
      }

      let textReplaced = false;
      for (const origBlock of originalBlocks) {
        const btype = _getBlockType(origBlock);

        if (btype === "text") {
          if (!textReplaced) {
            // First text block gets the (possibly compressed) content.
            blocks.push({ type: "text", text: msg.content });
            textReplaced = true;
          }
          // Subsequent text blocks are dropped — their content was
          // merged into msg.content during adaptIn.
        } else if (btype === "thinking") {
          // Pass thinking blocks through unchanged.
          if (_isPlainObject(origBlock)) {
            blocks.push({ ...(origBlock as Record<string, unknown>) });
          } else {
            blocks.push({
              type: "thinking",
              thinking: (_getNestedProp(origBlock, "thinking") as string) ?? "",
            });
          }
        } else if (btype === "tool_use") {
          const tcId = _getBlockProp(origBlock, "id");
          if (tcMap.has(tcId)) {
            const tc = tcMap.get(tcId)!;
            let inputData: unknown;
            try {
              inputData = JSON.parse(tc.function.arguments);
            } catch {
              inputData = { raw: tc.function.arguments };
            }
            blocks.push({
              type: "tool_use",
              id: tc.id,
              name: tc.function.name,
              input: inputData,
            });
          }
        } else {
          // Unknown / future block types — pass through unchanged.
          if (_isPlainObject(origBlock)) {
            blocks.push({ ...(origBlock as Record<string, unknown>) });
          } else if (typeof origBlock === "string") {
            blocks.push(origBlock as unknown as Record<string, unknown>);
          } else {
            blocks.push(btype ? { type: btype } : (origBlock as Record<string, unknown>));
          }
        }
      }

      // If no text block existed in originals but we have content, add it.
      if (!textReplaced && msg.content) {
        blocks.unshift({ type: "text", text: msg.content });
      }
    } else {
      // Fallback: reconstruct from scratch.
      if (msg.content) {
        blocks.push({ type: "text", text: msg.content });
      }

      if (msg.role === "assistant" && msg.toolCalls) {
        for (const tc of msg.toolCalls) {
          let inputData: unknown;
          try {
            inputData = JSON.parse(tc.function.arguments);
          } catch {
            inputData = { raw: tc.function.arguments };
          }
          blocks.push({
            type: "tool_use",
            id: tc.id,
            name: tc.function.name,
            input: inputData,
          });
        }
      }
    }

    result.push({ role: msg.role, content: blocks });

    // Re-nest tool results after assistant tool_use messages.
    if (msg.role === "assistant" && msg.toolCalls) {
      const trBlocks: Record<string, unknown>[] = [];
      for (const tc of msg.toolCalls) {
        if (toolResultMap.has(tc.id) && !consumedToolIds.has(tc.id)) {
          const trMsg = toolResultMap.get(tc.id)!;
          trBlocks.push({
            type: "tool_result",
            tool_use_id: tc.id,
            content: trMsg.content,
          });
          consumedToolIds.add(tc.id);
        }
      }
      if (trBlocks.length > 0) {
        result.push({ role: "user", content: trBlocks });
      }
    }
  }

  return result;
}

// ── Convenience function ───────────────────────────────────────────────────

export interface CompressAgentSdkOptions {
  llm?: MemoSiftLLMProvider | null;
  config?: Partial<MemoSiftConfig> | null;
  task?: string | null;
  ledger?: AnchorLedger | null;
  client?: unknown;
  model?: string;
  contextWindow?: import("../core/context-window.js").ContextWindowState | null;
}

/**
 * Compress Claude Agent SDK session messages end-to-end.
 *
 * Respects compaction boundaries (compact_boundary markers are
 * tagged _memosiftCompressed=true and skip re-compression).
 *
 * @param messages - Agent SDK message objects or session history dicts.
 * @param options - Configuration and provider options.
 * @returns Tuple of compressed Anthropic-format messages and compression report.
 */
export async function compressAgentSdkMessages(
  messages: unknown[],
  options?: CompressAgentSdkOptions,
): Promise<[Record<string, unknown>[], CompressResult["report"]]> {
  let provider = options?.llm ?? null;
  if (provider == null && options?.client != null) {
    provider = new ClaudeAgentLLMProvider(options.client, options.model ?? "claude-sonnet-4-6");
  }

  // Auto-resolve context window from model when not explicitly provided.
  let contextWindow = options?.contextWindow ?? null;
  if (contextWindow === null) {
    const { resolveContextWindow, estimateTokensHeuristic } = await import(
      "../core/context-window.js"
    );
    const model = options?.model ?? "claude-sonnet-4-6";
    const contentStrings = messages.map((m) => {
      if (typeof m === "object" && m !== null && "content" in m) {
        const c = (m as Record<string, unknown>).content;
        if (typeof c === "string") return c;
      }
      return "";
    });
    contextWindow = resolveContextWindow(null, model, estimateTokensHeuristic(contentStrings));
  }

  const memosiftMsgs = adaptIn(messages);
  const { messages: compressed, report } = await compress(memosiftMsgs, {
    llm: provider,
    config: options?.config,
    task: options?.task,
    ledger: options?.ledger,
    contextWindow,
  });
  return [adaptOut(compressed), report];
}

// ── Private helpers ────────────────────────────────────────────────────────

/**
 * Convert a plain dict message (from session history) to MemoSiftMessage(s).
 *
 * Returns a list because a single Anthropic-format user message with
 * tool_result blocks expands into multiple MemoSiftMessages (one per result).
 */
function _adaptDictMessage(msg: Record<string, unknown>): MemoSiftMessage[] {
  const role = (msg.role as string) ?? "user";
  const content = msg.content;

  if (!Array.isArray(content)) {
    return [
      createMessage(
        role as MemoSiftMessage["role"],
        typeof content === "string" ? content : String(content ?? ""),
      ),
    ];
  }

  // Preserve original blocks for lossless round-trip reconstruction.
  const originalBlocks = [...(content as unknown[])];

  const textParts: string[] = [];
  const toolCalls: ToolCall[] = [];
  const toolResults: Array<[string, string]> = [];

  for (const block of content as unknown[]) {
    if (!_isPlainObject(block)) continue;
    const b = block as Record<string, unknown>;

    if (b.type === "thinking") {
    } else if (b.type === "text") {
      textParts.push((b.text as string) ?? "");
    } else if (b.type === "tool_use") {
      toolCalls.push({
        id: (b.id as string) ?? "",
        type: "function",
        function: {
          name: (b.name as string) ?? "",
          arguments: JSON.stringify(b.input ?? {}),
        },
      });
    } else if (b.type === "tool_result") {
      let trContent = b.content;
      if (Array.isArray(trContent)) {
        trContent = (trContent as unknown[])
          .filter((item): item is Record<string, unknown> => _isPlainObject(item))
          .map((item) => (item.text as string) ?? String(item))
          .join(" ");
      }
      toolResults.push([(b.tool_use_id as string) ?? "", String(trContent ?? "")]);
    }
  }

  const result: MemoSiftMessage[] = [];
  const combinedText = textParts.length > 0 ? textParts.join("\n") : "";

  // Main message (assistant with tool_calls, or user with text).
  if (combinedText || toolCalls.length > 0) {
    result.push(
      createMessage(role as MemoSiftMessage["role"], combinedText, {
        toolCalls: toolCalls.length > 0 ? toolCalls : null,
        metadata: {
          _original_blocks: originalBlocks,
          _original_block_format: "claude_agent_sdk",
        },
      }),
    );
  }

  // Emit tool results as separate tool messages.
  for (const [toolUseId, trContent] of toolResults) {
    result.push(
      createMessage("tool", trContent, {
        toolCallId: toolUseId,
      }),
    );
  }

  // If we produced nothing, emit an empty message to preserve the role.
  if (result.length === 0) {
    result.push(createMessage(role as MemoSiftMessage["role"], ""));
  }

  return result;
}

/** Check if a value is a plain object (not an array, not null). */
function _isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Get constructor name from an object instance. */
function _getConstructorName(obj: unknown): string {
  if (obj == null) return "";
  if (typeof obj !== "object") return "";
  return (obj as { constructor?: { name?: string } }).constructor?.name ?? "";
}

/** Safely access a nested property on an unknown object. */
function _getNestedProp(obj: unknown, key: string): unknown {
  if (obj == null || typeof obj !== "object") return undefined;
  return (obj as Record<string, unknown>)[key];
}

/** Get the block type from either a typed object or a dict. */
function _getBlockType(block: unknown): string | null {
  if (_isPlainObject(block)) return (block as Record<string, unknown>).type as string | null;
  if (typeof block === "object" && block != null) {
    return (_getNestedProp(block, "type") as string) ?? null;
  }
  return null;
}

/** Get the text content from a block (typed object or dict). */
function _getBlockText(block: unknown): string {
  if (_isPlainObject(block)) return ((block as Record<string, unknown>).text as string) ?? "";
  return (_getNestedProp(block, "text") as string) ?? "";
}

/** Get a string property from a block (typed object or dict). */
function _getBlockProp(block: unknown, key: string): string {
  if (_isPlainObject(block)) return ((block as Record<string, unknown>)[key] as string) ?? "";
  return (_getNestedProp(block, key) as string) ?? "";
}
